core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java

```diff
@@ -0,0 +1,224 @@
+package jenkins.plugins;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import hudson.ClassicPluginStrategy;
+import hudson.PluginWrapper;
+import hudson.util.VersionNumber;
+import jenkins.util.java.JavaUtils;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Dedicated class to handle the logic related to so-called <em>detached plugins</em>.
+ *
+ * <p>Originally, some features were directly in Jenkins core. Over time, more and more features got extracted in dedicated plugins.
+ * Issue is: many plugins had started depending on these features, that now were not in the core anymore.
+ * So the chosen design strategy has been that the jenkins.war would embed these plugins, and automatically install them and mark them as optional dependencies.
+ * This way, older plugins would keep working without having to be modified.</p>
+ * <p>
+ * This code was originally moved from {@link ClassicPluginStrategy}.
+ *
+ * @since TODO
+ */
+@Restricted(NoExternalUse.class)
+public class DetachedPluginsUtil {
+    private static final Logger LOGGER = Logger.getLogger(DetachedPluginsUtil.class.getName());
+
+    /**
+     * Record of which plugins which removed from core and when.
+     */
+    @VisibleForTesting
+    static final List<DetachedPlugin> DETACHED_LIST;
+
+    /**
+     * Implicit dependencies that are known to be unnecessary and which must be cut out to prevent a dependency cycle among bundled plugins.
+     */
+    private static final Set<String> BREAK_CYCLES;
+
+    static {
+        try (InputStream is = ClassicPluginStrategy.class.getResourceAsStream("/jenkins/split-plugins.txt")) {
+            DETACHED_LIST = ImmutableList.copyOf(configLines(is).map(line -> {
+                String[] pieces = line.split(" ");
+
+                // defaults to Java 1.0 to install unconditionally if unspecified
+                return new DetachedPluginsUtil.DetachedPlugin(pieces[0],
+                                                              pieces[1] + ".*",
+                                                              pieces[2],
+                                                              pieces.length == 4 ? pieces[3] : "1.0");
```

Comment:
```suggestion
                                                              pieces.length >= 4 ? pieces[3] : "1.0");
```

Just in case there is a trailing data in the resource file

Reply:
I don't really see how this could happen. I mean, I would 100% agree for an external config file, but this file is internal and is very rarely changed. 
I guess if I was to do a change here, that would be more adding a test to check there's max 4 columns, and fail if not.

Reply:
Done as https://github.com/jenkinsci/jenkins/pull/3865/commits/a0e38ee9f8e93000177c609c8a5ea11050f0933e